src/utils/rootkeypackage_utils/src/rootkeypackage_download.c

```diff
@@ -0,0 +1,123 @@
+    ADUC_RootKeyPkgDownloaderInfo* downloaderInfo,
+    STRING_HANDLE* outRootKeyPackageDownloadedFile)
+{
+    ADUC_Result result = { .ResultCode = ADUC_GeneralResult_Failure, .ExtendedResultCode = 0 };
+
+    STRING_HANDLE targetDir = NULL;
+    STRING_HANDLE targetFileName = NULL;
+    STRING_HANDLE targetFilePath = NULL;
+    STRING_HANDLE targetUrl = NULL;
+
+    if (IsNullOrEmpty(rootKeyPkgUrl) || IsNullOrEmpty(workflowId) || downloaderInfo == NULL
+        || IsNullOrEmpty(downloaderInfo->name) || IsNullOrEmpty(downloaderInfo->downloadBaseDir)
+        || downloaderInfo->downloadFn == NULL || outRootKeyPackageDownloadedFile == NULL)
+    {
+        result.ExtendedResultCode = ADUC_ERC_UTILITIES_ROOTKEYPKG_DOWNLOAD_BADARG;
+        goto done;
+    }
+
+    if ((targetDir = STRING_construct_sprintf("%s/%s", downloaderInfo->downloadBaseDir, workflowId)) == NULL)
+    {
+        result.ExtendedResultCode = ADUC_ERC_NOMEM;
+        goto done;
+    }
+
+    if (ADUC_SystemUtils_MkDirRecursiveDefault(STRING_c_str(targetDir)) != 0)
+    {
+        result.ExtendedResultCode = ADUC_ERC_UTILITIES_ROOTKEYPKG_DOWNLOAD_MKDIR_DWNLD_FOLDER;
+        goto done;
+    }
+
+    if (IsNullOrEmpty(ADUC_ROOTKEY_PKG_URL_OVERRIDE))
+    {
+        targetUrl = STRING_construct(rootKeyPkgUrl);
+    }
+    else
+    {
+        targetUrl = STRING_construct(ADUC_ROOTKEY_PKG_URL_OVERRIDE);
+    }
+
+    if (targetUrl == NULL)
+    {
+        result.ExtendedResultCode = ADUC_ERC_NOMEM;
+        goto done;
+    }
+
+    // The filename should be the last path segment of the URL path.
+    result = ADUC_UrlUtils_GetLastPathSegmentOfUrl(STRING_c_str(targetUrl), &targetFileName);
+    if (IsAducResultCodeFailure(result.ResultCode) || IsNullOrEmpty(STRING_c_str(targetFileName)))
+    {
+        result.ResultCode = ADUC_GeneralResult_Failure;
+        result.ExtendedResultCode = ADUC_ERC_UTILITIES_ROOTKEYPKG_DOWNLOAD_URL_BAD_PATH;
+        goto done;
+    }
+
+    targetFilePath = STRING_construct_sprintf("%s/%s", STRING_c_str(targetDir), STRING_c_str(targetFileName));
+    if (targetFilePath == NULL)
+    {
+        result.ResultCode = ADUC_GeneralResult_Failure;
+        result.ExtendedResultCode = ADUC_ERC_NOMEM;
+        goto done;
+    }
+
+    // There is no hash to check so it is a forced download without lookup of
+    // existing file. Also, the package has self-referential integrity with
+    // signatures json array of signatures of the protected properties in the
+    // same json file, so download the file without hash verification.
+    Log_Debug("Attempting download of '%s' using '%s'", STRING_c_str(targetUrl), downloaderInfo->name);
+
+    result = (*(downloaderInfo->downloadFn))(STRING_c_str(targetUrl), STRING_c_str(targetFilePath));
+
+    if (IsAducResultCodeFailure(result.ResultCode))
+    {
+        Log_Error("Unable to download root key pkg.");
+        goto done;
+    }
+
+    *outRootKeyPackageDownloadedFile = targetFilePath;
```

Comment:
Same thing here. Won't outRootKeyPackageDownloadFile be set to a STRING_HANDLE that is deleted on line 116